File: app/src/main/java/com/github/ypicoleal/nad/ClickToSelectEditText.java
package com.github.ypicoleal.nad;

import android.content.Context;
import android.graphics.Canvas;
import android.support.v7.widget.AppCompatEditText;
import android.support.v7.widget.PopupMenu;
import android.util.AttributeSet;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import java.util.List;

public class ClickToSelectEditText<T extends Listable> extends AppCompatEditText {

    List<T> mItems;
    String[] mListableItems;
    CharSequence mHint;

    OnItemSelectedListener<T> onItemSelectedListener;

    public ClickToSelectEditText(Context context) {
        super(context);

        mHint = getHint();
    }

    public ClickToSelectEditText(Context context, AttributeSet attrs) {
        super(context, attrs);

        mHint = getHint();
    }

    public ClickToSelectEditText(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        mHint = getHint();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        setFocusable(false);
        setClickable(true);
        setLongClickable(false);
    }

    public void setItems(List<T> items) {
        this.mItems = items;
        this.mListableItems = new String[items.size()];

        int i = 0;

        for (T item : mItems) {
            mListableItems[i++] = item.getLabel();
        }

        configureOnClickListener();
    }

    private void configureOnClickListener() {
        final PopupMenu popup = new PopupMenu(this.getContext(), this);
        Menu menu = popup.getMenu();
        for (int i = 0; i < mItems.size(); i++) {
            menu.add(Menu.NONE, Menu.NONE, i + 1, mItems.get(i).getLabel());
        }

        popup.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {
                if (item.getOrder() > 0) {
                    int selectedIndex = item.getOrder() - 1;
                    setText(item.getTitle());
                    onItemSelectedListener.onItemSelectedListener(mItems.get(selectedIndex), selectedIndex);
                    return true;
                }
                return false;
            }
        });

        setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View view) {
                popup.show();
            }
        });
    }

    public void setOnItemSelectedListener(OnItemSelectedListener<T> onItemSelectedListener) {
        this.onItemSelectedListener = onItemSelectedListener;
    }

    interface OnItemSelectedListener<T> {
        void onItemSelectedListener(T item, int selectedIndex);
    }
}

File: app/src/main/java/com/dranilsaarias/nad/ClickToSelectEditText.kt
package com.dranilsaarias.nad

import android.content.Context
import android.graphics.Canvas
import android.support.v7.widget.AppCompatEditText
import android.support.v7.widget.PopupMenu
import android.util.AttributeSet
import android.view.Menu

class ClickToSelectEditText<T : Listable> : AppCompatEditText {

    internal lateinit var mItems: List<T>
    internal var mHint: CharSequence

    internal lateinit var onItemSelectedListener: OnItemSelectedListener<T>

    constructor(context: Context) : super(context) {

        mHint = hint
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {

        mHint = hint
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {

        mHint = hint
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        isFocusable = false
        isClickable = true
        isLongClickable = false
    }

    fun setItems(items: List<T>) {
        this.mItems = items

        configureOnClickListener()
    }

    private fun configureOnClickListener() {
        val popup = PopupMenu(this.context, this)
        val menu = popup.menu
        for (i in mItems.indices) {
            menu.add(Menu.NONE, Menu.NONE, i + 1, mItems[i].label)
        }

        popup.setOnMenuItemClickListener(PopupMenu.OnMenuItemClickListener { item ->
            if (item.order > 0) {
                val selectedIndex = item.order - 1
                setText(item.title)
                onItemSelectedListener.onItemSelectedListener(mItems[selectedIndex], selectedIndex)
                return@OnMenuItemClickListener true
            }
            false
        })

        setOnClickListener { popup.show() }
    }

    fun setOnItemSelectedListener(onItemSelectedListener: OnItemSelectedListener<T>) {
        this.onItemSelectedListener = onItemSelectedListener
    }

    interface OnItemSelectedListener<T> {
        fun onItemSelectedListener(item: T, selectedIndex: Int)
    }
}